akka-typed/src/main/scala/akka/typed/cluster/Cluster.scala

```diff
@@ -0,0 +1,314 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.typed.cluster
+
+import akka.actor.{ Address, ExtendedActorSystem }
+import akka.annotation.{ DoNotInherit, InternalApi }
+import akka.cluster.ClusterEvent.{ ClusterDomainEvent, CurrentClusterState, MemberEvent }
+import akka.cluster._
+import akka.cluster.protobuf.msg.ClusterMessages
+import akka.japi.Util
+import akka.typed.internal.adapter.ActorSystemAdapter
+import akka.typed.scaladsl.Actor
+import akka.typed.scaladsl.adapter._
+import akka.typed.{ ActorRef, ActorSystem, Extension, ExtensionId, PostStop, Terminated }
+
+import scala.collection.immutable
+import scala.reflect.ClassTag
+
+/**
+ * Messages for subscribing to changes in the cluster state
+ *
+ * Not intended for user extension.
+ */
+@DoNotInherit
+sealed trait ClusterStateSubscription
+
+object Subscribe {
+  def apply[A <: ClusterDomainEvent: ClassTag](subscriber: ActorRef[A]) =
+    new Subscribe[A](subscriber, implicitly[ClassTag[A]].runtimeClass.asInstanceOf[Class[A]])
+}
+
+/**
+ * Subscribe to cluster state changes. The initial state of the cluster will be sent as
+ * a "replay" of the subscribed events.
+ *
+ * If all you are interested in is to react on this node becoming a part of the cluster,
+ * see [[OnSelfUp]]
+ *
+ * @param subscriber A subscriber that will receive events until it is unsubscribed or stops
+ * @param eventClass The type of events to subscribe to, can be individual event types such as
+ *                   `ReachabilityEvent` or one of the common supertypes, such as `MemberEvent` to get
+ *                   all the subtypes of events.
+ */
+final case class Subscribe[A <: ClusterDomainEvent](
+  subscriber: ActorRef[A],
+  eventClass: Class[A]) extends ClusterStateSubscription
+
+/**
+ * Subscribe to this node being up, after sending this event the subscription is automatically
+ * cancelled. If the node is already up the event is also sent to the subscriber. If the node was up
+ * but is no more because it left or is leaving the cluster, no event is sent and the subscription
+ * request is ignored.
+ */
+final case class OnSelfUp(subscriber: ActorRef[SelfUp]) extends ClusterStateSubscription
+
+/**
+ * @param currentClusterState The cluster state snapshot from when the node became Up.
+ */
+final case class SelfUp(currentClusterState: CurrentClusterState)
+
+/**
+ * Subscribe to this node being removed from the cluster. If the node was already removed from the cluster
+ * when this subscription is created it will be responded to immediately from the subscriptions actor.
+ */
+final case class OnSelfRemoved(subscriber: ActorRef[SelfRemoved.type]) extends ClusterStateSubscription
+
+// TODO any additional data worth passing with this response?
+final case object SelfRemoved
```

Comment:
sometimes it has been requested to know how it was removed, if it was downed or left. We have that in `MemberRemoved` as `previousStatus`

Reply:
Good, I'll include that